MRDL->MRTK Tooltip UI control

cre8ivepark opened the pull request:
![mrtk_tooltip](https://user-images.githubusercontent.com/13754172/37489229-34b53d18-2854-11e8-9d0d-9835041f24b6.jpg)

Overview
---
Tooltips are usually used to convey a hint or extra information upon closer inspection of an object. ToolTip can be used to explain button inputs on the motion controllers or to label objects in the physical environment.

Demo video
---
https://gfycat.com/WarmOblongBilby

Example scene
---
This example scene demonstrates an implementation of the ToolTip user interface element. 

[TooltipExamples.unity](https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/Feature_UX_Tooltip/Assets/MixedRealityToolkit-Examples/UX/Scenes/TooltipExamples.unity)
![mrtk_tooltipexamplescene](https://user-images.githubusercontent.com/13754172/37489230-34cbb052-2854-11e8-947a-b6c00c239053.jpg)

Technical details
---
https://github.com/Microsoft/MixedRealityToolkit-Unity/blob/Feature_UX_Tooltip/Assets/MixedRealityToolkit-Examples/UX/Readme/README_TooltipExample.md

DDReaper commented:
Quick Q @cre8ivepark , does the tooltip example also include Tooltips on at least one of the controllers? (as shown in the image above)

cre8ivepark commented:
Yes I have included tooltip groups for both left/right motion controllers. They will be attached when the controllers are detected. Please take a look at the example scene.

cre8ivepark commented:
@SimonDarksideJ You can also check out the demo video 
https://gfycat.com/WarmOblongBilby

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+    /// </summary>
+    public class ToolTipConnector : MonoBehaviour
+    {
+        public enum FollowTypeEnum
```
Can we separate out these Enums and if possible make them more generic? (one for each enum).  Also don't suffix them with Enum, as this is determined by its type.
Seems these are fairly common UX style enums that should be standardised and reused.

If these really are ONLY for this component, then they should be private.

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+    /// </summary>
+    public class ToolTipSpawner : MonoBehaviour , IInputHandler, IPointerSpecificFocusable
+    {
+        public enum VanishBehaviorEnum
```
Same comment as with Tooltip Connector, either generalise these for all UX as separate public scripts or make them private. As well as removing enum suffix

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+            hasFocus = true;
+            if (toolTip == null || !toolTip.gameObject.activeSelf)
+            {
+                switch (AppearBehavior)
```
if there is only one case, then just use IF not switch.

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+            tappedTime = Time.unscaledTime;
+            if (toolTip == null || !toolTip.gameObject.activeSelf)
+            {
+                switch (AppearBehavior)
```
Same as above, if only one case, use IF not switch.

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+
+                }
+                //check whether we're suppose to disappear
+                switch (VanishBehavior)
```
Good use of Switch.  Having multiple cases and a defined default.

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+
+            switch (AppearBehavior)
+            {
+                case AppearBehaviorEnum.AppearOnFocusEnter:
```
Same as above, use IF with switches with only one case and no default.

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipUtility.cs:
```diff
@@ -0,0 +1,117 @@
+    const int NumPivotLocations = 8;
+
+    [Serializable]
+    public enum AttachPointTypeEnum
```
Same Enum comments as before:
* Separate out to unique cs file for public and make generic
* Remove enum suffix

SimonDarksideJ commented:
Also noted one issue with z fighting between two tooltips as shown here. Not sure if you can, but can we try and solve this?
![image](https://user-images.githubusercontent.com/1793042/37555265-d88491b2-29dc-11e8-895a-bcf64454925d.png)

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Models/Materials/Default.mat:
```diff
@@ -0,0 +1,76 @@
+%YAML 1.1
```
Don't we already have some sort of default material?

StephenHodgson replied:
Ah I see, this was an imported material. Please disable material import on the model you imported.

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Models/TheModule.fbx.meta:
```diff
@@ -0,0 +1,242 @@
+    400006: BottomRightBody
+    400008: CenterBackLeg
+    400010: CenterFrontLeg
+    400012: Cylinder01
```
Could we properly name all the meshes in this object?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Models/TheModule.fbx.meta:
```diff
@@ -0,0 +1,242 @@
+    400052: TopGoldDial
+    400054: TopLeftBody
+    400056: TwirlerConnector
+    400058: Camera
```
the fbx has a camera embedded in it?

StephenHodgson commented on Assets/MixedRealityToolkit-Examples/UX/Models/TheModule.fbx.meta:
```diff
@@ -0,0 +1,242 @@
+  materials:
+    importMaterials: 1
+    materialName: 1
+    materialSearch: 2
```
Please disable material import.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackground.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+            if (toolTip == null)
+            {
+                toolTip = gameObject.GetComponent<ToolTip>();
```
use `gameObject.EnsureComponent<Tooltip>();` and add the `ToolTip` required component class attribute.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackground.cs:
```diff
@@ -0,0 +1,43 @@
+        protected abstract void ScaleToFitContent();
+
+        [SerializeField]
+        protected ToolTip toolTip;
```
fields and properties should be declared at the top of the class

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackground.cs:
```diff
@@ -0,0 +1,43 @@
+﻿//
```
use standard header

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+        public const float MaxDistortionCorrection = 5f;
+        public const float MaxRotationCorrection = 5f;
+
+        [Header("Transform targets")]
```
let's move the attribute declaration down below the summary

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+        /// Which transforms to use for each type of distortion
+        /// See the ToolTipBalloon prefab for an example of which transforms to target
+        /// </summary>
+        public Transform PositionTarget;
```
private fields with public properties if access is required outside the class.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+        {
+            base.OnEnable();
+            lastPosition = PositionTarget.position;
+            inertialContentBounds = new Bounds(Vector3.zero, Vector3.one);
```
this default bounds could be declared in a private field and reused each time it's needed to reset values.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+
+            // Get the size of the mesh and use this to adjust the local content size on the x / y axis
+            // This will accomodate meshes that aren't built to 1,1 scale
+            Bounds meshBounds = BackgroundRenderer.GetComponent<MeshFilter>().sharedMesh.bounds;
```
If the background renderer doesn't change often we can probably cache the `MeshFilter` so we don't have to get the component each time we scale this.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+            toolTip.AttachPointPosition = AttachPointOffset.position;
+
+            lastPosition = currentPosition;
+            
```
lots of extra space.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundBlob.cs:
```diff
@@ -0,0 +1,200 @@
+        private Vector3 distortion;
+        private Vector3 rotation;
+        private Bounds localContentBounds;
+        private Bounds inertialContentBounds;
```
move these to the other private fields declared at the top of the class.

johnppella replied:
fields moved.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundCorners.cs:
```diff
@@ -0,0 +1,105 @@
+            Vector3 botRight = new Vector3(localContentSize.x + localContentOffset.x, -localContentSize.y + localContentOffset.y, localContentOffset.x);
+            Vector3 botLeft = new Vector3(-localContentSize.x + localContentOffset.x, -localContentSize.y + localContentOffset.y, localContentOffset.x);
+            if (cornerTopLeft != null)
+                cornerTopLeft.localPosition = topLeft;
```
nit: missing braces

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundCorners.cs:
```diff
@@ -0,0 +1,105 @@
+            Vector3 globalScale = Vector3.one;
+            switch (scaleMode) {
+                case ScaleModeEnum.World:
+                default:
```
let's make the default throw the out of bounds exception.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipBackgroundMesh.cs:
```diff
@@ -0,0 +1,102 @@
+        /// <summary>
+        /// The z depth of the mesh background
+        /// </summary>
+        public float Depth = 1f;
```
private fields with public properties if access is required.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+        
+        public GameObject Target;
+
+        public FollowTypeEnum FollowType = FollowTypeEnum.AnchorOnly;
```
private fields with public properties if access is required.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+        {
+            if (toolTip == null)
+            {
+                toolTip = GetComponent<ToolTip>();
```
if the `Tooltip` is required for this class and gameobject then be sure to add the required component attribute and use `gameObject.EnsureComponent();`

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+        private void OnEnable()
+        {
+            if (!FindToolTip())
+                return;
```
missing braces

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+            {
+                toolTip = GetComponent<ToolTip>();
+            }
+            if (toolTip == null)
```
`return toolTip != null;`

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+            switch (FollowType)
+            {
+                case FollowTypeEnum.AnchorOnly:
+                default:
```
default case should throw array out of bounds exception

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+                case FollowTypeEnum.PositionOnly:
+                    // Move the entire tooltip transform while maintaining the anchor position offset
+                    toolTip.transform.position = Target.transform.position;
+                    switch (PivotMode)
```
let's try to put the nested switch statements in their own methods to prevent confusion and reduce nesting.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+                            switch (PivotDirectionOrient)
+                            {
+                                case OrientTypeEnum.OrientToCamera:
+                                    relativeTo = Camera.main.transform;//Veil.Instance.HeadTransform;
```
use `CameraCache.Main.transform`

StephenHodgson replied:
also remove comment

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipConnector.cs:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        [SerializeField]
+        private ToolTip toolTip;
```
fields should be declared at the top of classes

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+        }
+
+        [Range(0f, 5f)]
+        public float AppearDelay = 0.0f;
```
private fields with public properties if access outside the class is required

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+                toolTip.ShowConnector = showConnector;
+                toolTip.transform.position = transform.position;
+                toolTip.transform.parent = transform;
+                toolTip.ContentParentTransform.localScale = new Vector3(0.182f, 0.028f, 1.0f);
```
we  can probably cache this vector value in a serialized private field.  What's with the magic numbers anyway?  Is this something a dev might want to customize?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+                    case VanishBehaviorEnum.VanishOnTap:
+                        if (tappedTime != tappedTimeOnStart)
+                        {
+                            toolTip.gameObject.SetActive(false);
```
instead of just popping the tooltip on and off, should we animate them?

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipSpawner.cs:
```diff
@@ -0,0 +1,263 @@
+        private float focusExitTime = 0f;
+        private float tappedTime = 0f;
+        private bool hasFocus;
+        private ToolTip toolTip;
```
fields should be declared at the top of the class

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipUtility.cs:
```diff
@@ -0,0 +1,117 @@
+    /// meets the label.
+    /// </summary>
+
+    const int NumPivotLocations = 8;
```
missing access modifier and there's an extra line space here.

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipUtility.cs:
```diff
@@ -0,0 +1,117 @@
+
+    const int NumPivotLocations = 8;
+
+    [Serializable]
```
not sure this attribute is valid

StephenHodgson commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipUtility.cs:
```diff
@@ -0,0 +1,117 @@
+        float nearDist = Mathf.Infinity;
+
+        if (localPivotPositions == null || localPivotPositions.Length < NumPivotLocations)
+            return nearPivot;
```
missing braces

cre8ivepark commented:
- Applied John's updates that address Simon's feedback.
- Merged Dev branch. Resolved conflicts.
- Removed fbx imported materials, replaced with MRTK/Standard shader based materials. 

![2018-03-19 17_08_23-unity 2017 2 1p2 64bit - tooltipexamples unity - johnnew - universal windows p](https://user-images.githubusercontent.com/13754172/37629015-87d3c9ea-2b99-11e8-8fae-ecf990a20470.png)

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/AppBar/AppBar.cs:
```diff
@@ -9,7 +9,9 @@
-using UnityEngine.XR.WSA.Input;
+
+
+//using UnityEngine.XR.WSA.Input;
```
Don't leave commented out code in the file. Just delete it. (git is your friend)

SimonDarksideJ commented on Assets/MixedRealityToolkit/UX/Scripts/Tooltip/ToolTipUtility.cs:
```diff
@@ -0,0 +1,120 @@
+    const int NumPivotLocations = 8;
+
+    [Serializable]
+    public enum AttachPointType
```
Public enums should be in separate class files

cre8ivepark commented:
Applied John's updates, refined the motion controller example.
